### PR TITLE
[OneEmbedding] support clip by global norm

### PR DESCRIPTION
**Pull request by guo-ran**
#### pytorch中clip grad norm的具体实现

接口：https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html

```
torch.nn.utils.clip_grad_norm_(parameters, max_norm, norm_type=2.0, error_if_nonfinite=False)
```
norm_type接收数字和inf

```
    if norm_type == inf:
        norms = [p.grad.detach().abs().max().to(device) for p in parameters]
        total_norm = norms[0] if len(norms) == 1 else torch.max(torch.stack(norms))
    else:
        total_norm = torch.norm(torch.stack([torch.norm(p.grad.detach(), norm_type).to(device) for p in parameters]), norm_type)
    
    clip_coef = max_norm / (total_norm + 1e-6)
    clip_coef_clamped = torch.clamp(clip_coef, max=1.0)
    for p in parameters:
        p.grad.detach().mul_(clip_coef_clamped.to(p.grad.device))
```

计算过程是:

如果是inf，就对每个梯度求reduce_abs_max，如果有多个结果，就stack起来再求一次max，得到total_norm.

如果是数字，就对每个梯度求pow(reduce_sum_pow(grad, norm_type), 1/norm_type)，如果norm_type是2，就是平方和开根号。如果有多个结果，就stack起来再求一次norm，得到total_norm。

求到total_norm之后，再计算求得clip_coef_clamped 乘到每个grad上。

#### OneFlow中clip grad norm的具体实现
目前配置方式是

```
        params = [
            {
                "params": xxmodule.parameters(),
                "clip_grad_max_norm": 0.5,
                "clip_grad_norm_type": norm_type,
            },
        ]
        self.add_optimizer(flow.optim.SGD(params, lr=0.1, momentum=0.0))
```

代码在https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/job_rewriter/autograd.cpp#L697

如果norm_type是inf就执行GlobalAbsMaxMin，只是norm_type为正的时候里面是max，负的时候里面是min。否则就执行GlobalNorm。

GlobalAbsMaxMin的逻辑是对每个grad，如果有split，就先执行local_multi_reduce_max然后执行reduce_max，分两步得到grads的全局max值。如果有多个就接一个stack，再执行一次reduce_max，得到total_norm_lbn。

GlobalNorm的逻辑是对多个grad执行multi_reduce_sum_pow_abs，然后接一个add，然后接一个scalar_pow 1/p，p如果是2，可以理解为平方和再开根号。

这两处代码都涉及到多个grad如果placement不同，就把求total norm的op放在cpu 0上，如果placement相同，就放在相同的placement上。

在得到total_norm_lbn后先对total_norm_lbn接一个scalar_add 1e-6操作，然后接一个reciprocal_no_nan操作求倒数，然后接一个scalar_mul操作乘max_norm，再接一个clip_by_scalar_max 操作clip 1.0，得到coeff_lbn，会乘在每个grad上。

这几步实际上做的就是 clamp(max_norm / (total_norm + 1e-6), max=1.0) 。和torch的是等价的。

#### OneEmbedding中的实现
以norm_type为2举例，在autograd中对所有的grad执行平方和开根号得到total_norm:

```
total_norm = sum(reduce_sum(grad^2))^1/2
coff = clip(max_norm / (total_norm + 1e-6), 1)
```

最后对所有的grad做scale: `grad = grad * coff  `

在OneEmbedding的pass中，对于embedding_grad，需要和它所在的OptimizerConf下的所有grad一起求得total_norm和coff，最后做scale。若embedding_grad所属的OptimizerConf下没有已求出的total_norm，则embedding_grad自己求total_norm和coff并作用于梯度。

实现过程是：

1、先找出所有的shadow_op2embedding_grad，并遍历OptimizerConf，过滤出属于本OptimizerConf下shadow_op2embedding_grad

2、对这些embedding_grad 求norm(若norm_type为2，是求平方和)，再查找TotalNormState：

(1) 若能找到，说明当前OptimizerConf下有其他grad，从TotalNormState中取出total_norm_lbn，将embedding_grad求得的norm和total_norm_lbn结合计算得到new_total_norm，并用new_total_norm 更新原来total_norm_lbn的consumer。consumer更新后，coff也是新的total_norm计算得到的结果，再用coff scale embedding_grad即可。

(2)若找不到，说明当前OptimizerConf下没有其他grad，则将embedding_grads求得的norm作为total_norm，并接这些op得到coff，`coff = clip(max_norm / (total_norm + 1e-6), 1)`，用coff scale梯度。

需要考虑的点：

1、相同OptimizerConf下的grad一起做clip，不同OptimizerConf下的grad不一起做clip，因此在autograd中，每个OptimizerConf下的grad求得一份TotalNormState，用variable_op_name索引它。在OneEmbedding的pass中，根据shadow_op_name找到相同OptimizerConf下的任意一个variable_op_name，拿到它索引的TotalNormState。

2、原来的实现中在amp下，若ONEFLOW_ONE_EMBEDDING_GRADIENT_SHUFFLE_USE_FP16为true(默认为true)，gradient shuffle会用fp16类型，得到的grad也是fp16类型，会将cast、loss_instance_num scale、loss scale都fuse进update op里。在有clip grad存在下，就不能fuse进update op了，需要对grad先cast成float，再scale，再接update op。即：

(1) 如果有clip grad，cast、scale和scale_by_tensor就不fuse到update上，grad就是float类型
    gradient shuffle grad->cast h2f -> scale_by_tensor->scale-> new_grad
    new_grad 接clip grad
    new_grad 接count not finite
    new_grad 接 update
(2) 如果没有clip grad，就fuse到update上
    gradient shuffle grad  接  count not finite
    gradient shuffle grad  接 update
    loss instance num和static loss scale的scale 作为update的属性
    dynamic loss scale 是down_scale_ptr传给update.

注意事项：
目前embedding_grad是gradient_shuffle的结果，有部分是有效值，有部分是无效值，无效值部分是填0，所以如果norm_type是-inf，AbsReduceMin的结果就永远是0，因此暂时不允许norm_type 为-inf的情况(目前torch的clip grad中norm_type也没有-inf)。

### CODE REVIEW

File: oneflow/core/job_rewriter/replace_embedding_ops_pass.cpp [hunk 1]
```diff
@@ -537,6 +588,292 @@ void UpdateConsumerOpConf(const OpNode* consumer, const LogicalBlobId& out,
...
+                                 const ParallelConf& parallel_conf,
+                                 const int64_t scope_symbol_id) -> std::string {
+    if (std::isinf(norm_type) && norm_type > 0) {
+      return GlobalAbsMaxMin(job_builder, shadow_op_name2grad_lbn, norm_type, total_norm_lbn, true,
+                             embedding_parallel_conf, embedding_scope_symbol_id, parallel_conf,
+                             scope_symbol_id);
+    } else if (std::isinf(norm_type) && norm_type < 0) {
+      UNIMPLEMENTED()
+          << "one_embedding gradient's invalid values set to 0, so not support abs_reduce_min.";
+      return GlobalAbsMaxMin(job_builder, shadow_op_name2grad_lbn, norm_type, total_norm_lbn, false,
```
Comment (MARD1NO) [inline oneflow/core/job_rewriter/replace_embedding_ops_pass.cpp:785]: 前面UNIMPLEMENTED了，那这里是不是就没必要了

**Review by MARD1NO — APPROVED** [review]

File: oneflow/core/job_rewriter/replace_embedding_ops_pass.cpp (hunk 1 above)
Comment (guo-ran) [reply oneflow/core/job_rewriter/replace_embedding_ops_pass.cpp:785]: 这里未来解决无效值的问题后这部分结果就是正确的，就可以只移除UNIMPLEMENTED了，所以我就保留了

**Review by liujuncheng — APPROVED** [review]
